### ISSUE TITLE
Surprisingly large number of allocations

### ISSUE THREAD

**Issue by NiklasGustafsson** [FluxML/Zygote.jl#202]
The following code produces a surprising number of allocations when differentiated with Zygote:

```
import Pkg; Pkg.activate(".")

using Zygote

const Mat = AbstractMatrix{Float64}

derivative(f::Function, x...) = Zygote.gradient(f, x...)[1]

loss(params::Mat) = 0.0

weights = rand(128,784)

function train(loss, weights::Mat, iterations=1000)

    for i = 1 : iterations
        grads = derivative(loss, weights)
    end

  return weights
end

@time w = train(loss, weights)
```

The output I get on Windows is:
```
> julia .\repro.jl                                                                                                                 
4.546076 seconds (12.25 M allocations: 614.791 MiB, 4.48% gc time)
```

**Comment by gdkrmr** [comment]
this looks like you are measuring compile time. Try running the last line two times and see the difference in allocations/execution time.

**Comment by NiklasGustafsson** [comment]
Yes, it does. Thank you for pointing it out. In the original, not-pared-down repro, I did not, but this is clearly a bogus measurement. With a proper warm-up run, there are only 9 allocations per iteration. Not sure why it should need any allocations, but it's at least not excessive.

I'll keep the issue open and post the larger repro in my next comment.

**Comment by NiklasGustafsson** [comment]
Here is the larger repro -- a simple fully connected network with an MSE-based loss function. Softmax layer removed, but activations still in place for each layer. I removed the optimizer, so that it's just Zygote that is being measured. Almost 19K allocations per iteration:

```
using Zygote

const Vec = AbstractVector{Float64}
const Mat = AbstractMatrix{Float64}

struct DenseParams
    W :: Mat
    b :: Vec
end

lrelu(x, α=0.1) = max(x, α * x)

sqnorm(x) = sum(x .* x);
mse(x::Mat, y::Mat) = sqnorm(x .- y)

dnn(weights::Array{DenseParams}, x::Mat) = foldr((p,input) -> lrelu.(p.W * input .+ p.b), weights, init=x)    

loss(params, examples, labels) = mse(dnn(params,examples), labels)

derivative(f::Function, x...) = Zygote.gradient(f, x...)[1]

function train(loss, w, inputs, labels, iterations=1000)

  for i = 1 : iterations
    grads = derivative(loss, w, inputs, labels)
  end

  return w
end

inputs = rand(784,33)
labels = floor.(rand(10,33))

weights = [
        DenseParams(rand(10,32), zeros(10)), 
        DenseParams(rand(32,128), zeros(32)), 
        DenseParams(rand(128,784), zeros(128))]

train(loss, weights, inputs, labels, 1)

@time w = train(loss, weights, inputs, labels)
```
The output:

```
> julia .\repro.jl
2.495737 seconds (18.93 M allocations: 1.931 GiB, 9.43% gc time)

```

**Comment by gdkrmr** [comment]
I recommend you read the manual:

https://docs.julialang.org/en/v1/manual/performance-tips/#Measure-performance-with-[@time](@ref)-and-pay-attention-to-memory-allocation-1

and checkout https://github.com/JuliaCI/BenchmarkTools.jl

**Comment by NiklasGustafsson** [comment]
Measuring the code I control, i.e. the 'loss' function itself, not its gradient, shows that it's actually quite efficient from an allocation perspective -- 36 allocations per iteration, which doesn't seem outrageous.

By not using AbstractVector/Matrix in the definitions of Vec and Mat, it goes from 36 to 24. The gradient calculation remains at 19K per iteration.

**Comment by staticfloat** [comment]
I looked into this with @NiklasGustafsson today, and it appears that this is due to broadcasting; something in the inverse broadcast machinery is causing very large numbers of allocations.  MWE:

```julia
using Zygote
Zygote.refresh()

x = randn(100,100)
w = randn(100,100)

model(w) = sum(identity.(x + w))

# Run forward model
y, back = forward(model, w)
@info("Forward timing:")
@time forward(model, w)

# Run backwards model
back(y)
@info("Backward timing:")
@time back(y);
```

On `v1.2-rc1`, I get the following timings:
```
[ Info: Forward timing:
  0.000085 seconds (163 allocations: 241.906 KiB)
[ Info: Backward timing:
  0.000543 seconds (30.08 k allocations: 705.641 KiB)
```

Removing the `.(` to define `model(w)` as `sum(identity(x + w))`, improves the situation drastically:
```
[ Info: Forward timing:
  0.000087 seconds (147 allocations: 85.203 KiB)
[ Info: Backward timing:
  0.000124 seconds (93 allocations: 81.359 KiB)
```

Intriguingly, using operations such as `.+` and `.*` do not seem to impose this allocation penalty, it is only `.(` which makes me wonder if perhaps there's some missing Zygote adjoint definition for a general `Broadcasted` or `materialize` or something.

**Comment by staticfloat** [comment]
Digging into Zygote a bit, I believe we may be hitting this: https://github.com/FluxML/Zygote.jl/blob/v0.3.1/src/lib/broadcast.jl#L73-L84

@MikeInnes is this kind of poor performance expected for broadcasting?  We could, of course, write a custom `@adjoint` for the `lrelu()` call in Niklas' code above, but that's a workaround rather than a true solution to poor broadcasting performance (which we would all like, I think).  A quick search of the issues/PRs doesn't show any discussion about the linked note on poor performance, could you write up what the root problem is so that we can try to quantify the issue?

**Comment by MikeInnes** [comment]
Broadcasting is hard. I basically don't have any great insights beyond the comments in that code; Base's machinery appears to be right on the edge of what the compiler can handle and adding any complexity sends us off of a cliff, but I don't know exactly what goes wrong there.

Aside from digging into reverse mode to fix this, one option would be to heuristically dispatch to fused forward mode for real -> real functions, which would catch cases like this. We could also have an explicit "use fused forward mode" annotation.

A simpler workaround is to write `lrelu` Python-style, e.g. `lrelu(xs) = max.(xs, α .* xs)`, explicitly broadcasting the inner operations rather than the whole `relu`; then Zygote will catch those operations explicitly. Alternatively perhaps just use NNlib's `leakyrelu` and we can make sure it's fast.

**Comment by CarloLucibello** [comment]
The allocation explosion is gone; the same loop now runs in ~3ms with no per-iteration blowup. Closing as fixed.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
